terraform-post-packer/main.tf

```
@@ -29,7 +29,7 @@ data "aws_ami_ids" "historical_amis" {
   filter {
     name = "name"
     values = [
-      "example-hvm-*-x86_64-ebs",
+      "windows-commando-hvm-*-x86_64-ebs",
```

If you remove the "hvm" bit, don't forget to also do it here:
```suggestion
      "windows-commando-*-x86_64-ebs",
```

---

@dav3r when I try to remove the hvm like you mentioned here. I get this precommit error:

```
Terraform fmt............................................................Failed
- hook id: terraform_fmt
- exit code: 1

/Users/mostafaabdelbaky/.cache/pre-commit/repoyjmurj0f/hooks/terraform_fmt.sh: line 9: realpath: command not found
/Users/mostafaabdelbaky/.cache/pre-commit/repoyjmurj0f/hooks/terraform_fmt.sh: line 11: ./_common.sh: No such file or directory

Terraform validate.......................................................Failed
- hook id: terraform_validate
- exit code: 1

/Users/mostafaabdelbaky/.cache/pre-commit/repoyjmurj0f/hooks/terraform_validate.sh: line 9: realpath: command not found
/Users/mostafaabdelbaky/.cache/pre-commit/repoyjmurj0f/hooks/terraform_validate.sh: line 11: ./_common.sh: No such file or directory```

I've been trying to debug the issue in the meantime

---

@itsmostafa Are you sure your virtual environment is set up correctly?  You can try running `setup-env -f` to create a clean virtual env and see if that helps.

---

@itsmostafa Were you running pre-commit locally before? That looks like you're missing the `realpath` binary on your system. If you're on a Mac you'd need to have `coreutils` installed with [brew](https://github.com/Homebrew/brew). You could try `command -v realpath` to check on the binary.

---

@dav3r @mcdonnnj I had my precommit setup locally from when I first started working on this project. It was precommitting successfully before. I didnt have realpath though so I installed it via brew and now I'm getting some new errors

---

```
First I uninstalled and reinstalled pre-commit.

Terraform fmt............................................................Failed
- hook id: terraform_fmt
- exit code: 2

/Users/mostafaabdelbaky/.cache/pre-commit/repoyjmurj0f/hooks/_common.sh: line 30: declare: -g: invalid option
declare: usage: declare [-afFirtx] [-p] [name[=value] ...]

Terraform validate.......................................................Failed
- hook id: terraform_validate
- exit code: 1

Validation failed: terraform-post-packer
╷
│ Warning: Empty provider configuration blocks are not required
│ 
│   on .terraform/modules/ami_launch_permission/providers.tf line 3:
│    3: provider "aws" {
│ 
│ Remove the aws.master provider block from module.ami_launch_permission. Add
│ aws.master to the list of configuration_aliases for aws in
│ required_providers to define the provider configuration name.
╵
╷
│ Error: Missing required provider
│ 
│ This configuration requires provider registry.terraform.io/hashicorp/aws,
│ but that provider isn't available. You may be able to install it
│ automatically by running:
│   terraform init
```

I also tried running terraform init in the `terraform-post-packer` directory it returned this error:

```
terraform-post-packer git:(first-commits) ✗ sudo terraform init
Initializing modules...

Initializing the backend...
│ Warning: Empty provider configuration blocks are not required
│ 
│   on .terraform/modules/ami_launch_permission/providers.tf line 3:
│    3: provider "aws" {
│ 
│ Remove the aws.master provider block from module.ami_launch_permission. Add aws.master to
│ the list of configuration_aliases for aws in required_providers to define the provider
│ configuration name.
╵

╷
│ Error: error configuring S3 Backend: no valid credential sources for S3 Backend found.
│ 
│ Please see https://www.terraform.io/docs/language/settings/backends/s3.html
│ for more information about providing credentials.
│ 
│ Error: NoCredentialProviders: no valid providers in chain. Deprecated.
│       For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

---

@itsmostafa Have you gotten this error resolved yet?

---

@dav3r i've only gotten the `Terraform fmt` resolved with this commit [55e36f9](https://github.com/cisagov/windows-commando-packer/pull/10/commits/55e36f971332bbebf03664e47978ecdf4ff421ab)

`Terraform validate` doesn't fail until I try to change line 32 by removing `hvm` from the name. Then it fails with the error message above. As of now its good unless you need me to change that name.

---

With line 32 as it is now (`"windows-commando-hvm-*-x86_64-ebs",`), are you able to successfully run `terraform init` in the `terraform-post-packer` directory?

---

Hey @dav3r i just ran terraform init in the `terraform-post-packer` directory. This is the error I got:

```
Initializing modules...
Downloading git::https://github.com/cisagov/ami-launch-permission-tf-module.git for ami_launch_permission...
- ami_launch_permission in .terraform/modules/ami_launch_permission

Initializing the backend...
╷
│ Warning: Empty provider configuration blocks are not required
│ 
│   on .terraform/modules/ami_launch_permission/providers.tf line 3:
│    3: provider "aws" {
│ 
│ Remove the aws.master provider block from module.ami_launch_permission. Add aws.master to the list of
│ configuration_aliases for aws in required_providers to define the provider configuration name.
╵

╷
│ Error: error configuring S3 Backend: no valid credential sources for S3 Backend found.
│ 
│ Please see https://www.terraform.io/docs/language/settings/backends/s3.html
│ for more information about providing credentials.
│ 
│ Error: NoCredentialProviders: no valid providers in chain. Deprecated.
│       For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

---

@dav3r i ran `terraform-init` in the `terraform-post-packer` directory and this is the error i got:

```
terraform-post-packer git:(first-commits) terraform init
Initializing modules...

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.
Error refreshing state: AccessDenied: Access Denied
        status code: 403, request id: 19F68AXKMH6WAYHV, host id: WeKKLo91GYZylkbXL8GcJbO89bMa+J9bubQ0ZQXjTCJGP7GQRP229+ajgw1PX5gW9HA2K7eDo2Q=
```

---

@itsmostafa I don't think we ever gave you access to assume the two roles that you need for the post-packer Terraform to work.

@jsf9k or @mcdonnnj - I don't remember what we decided here.  Were we planning to give @itsmostafa access or were we just going to run the post-packer Terraform for him?

---

I think we can just run it for him and see if that becomes onerous for us or causes long delays for him.

---

@itsmostafa I just attempted to run this, but it doesn't look like you have created any AMIs yet (i.e. you haven't created any "prerelease" releases), so there is nothing for this post-packer Terraform to do just yet.

Note that this post-packer TF will not work as intended until the AMI names are in alignment:
https://github.com/cisagov/windows-commando-packer/blob/19dc3a3d8b268d0ca52347b84411bae2631ca8d4/src/packer.pkr.hcl#L71

https://github.com/cisagov/windows-commando-packer/blob/19dc3a3d8b268d0ca52347b84411bae2631ca8d4/terraform-post-packer/main.tf#L27-L33

---

@dav3r if that's the case, how do i get past this pre-commit error for post-packer?

```
Terraform validate.......................................................Failed
- hook id: terraform_validate
- exit code: 1

Validation failed: terraform-post-packer
╷
│ Warning: Empty provider configuration blocks are not required
│ 
│   on .terraform/modules/ami_launch_permission/providers.tf line 3:
│    3: provider "aws" {
│ 
│ Remove the aws.master provider block from module.ami_launch_permission. Add
│ aws.master to the list of configuration_aliases for aws in
│ required_providers to define the provider configuration name.
╵
╷
│ Error: Missing required provider
│ 
│ This configuration requires provider registry.terraform.io/hashicorp/aws,
│ but that provider isn't available. You may be able to install it
│ automatically by running:
│   terraform init
```

I tried running `terraform` init like it suggested but then i get the access denied error again.
Creating a prerelease fails for the same reason as well. Should I bypass precommit for now?

---

Yes, you can skip precommit for that one commit.  The prerelease GH Action should be using the build user creds that @jsf9k added as GH Secrets, so I'm not quite sure why it would be getting that same error.

---

@dav3r is this resolved?

---

I am resolving this conversation.  We can discuss AMI creation over in [this new thread](https://github.com/cisagov/windows-commando-packer/pull/10#discussion_r849827340).